### ISSUE TITLE
Cyclic dependency

### ISSUE THREAD

**Issue by C-Otto** [TracksApp/tracks#1754]
I created a few new tasks in a fresh project and then connected those using drag-n-drop. By accident, I created a dependency of taskA to taskB, although I created the reverse dependency just before that. I remember seeing some red error message flash up (which I ignored/did not read). I assumed the dependency was not added, as it would create a cycle.

However, according to my database, there is the cyclic dependency:

```
tracks=# select * from dependencies where successor_id in (5063,5064);
  id  | successor_id | predecessor_id | relationship_type 
------+--------------+----------------+-------------------
 1098 |         5064 |           5063 | 
 1099 |         5063 |           5064 | 
```

Running with PostgreSQL and on f81efebd8ce18bbca9abd567d345edbc22e80061.

**Comment by C-Otto** [comment]
Log file: http://paste.debian.net/144057/

**Comment by C-Otto** [comment]
See #1401

**Comment by C-Otto** [comment]
The cycle is introduced in `@predecessor_array << t` (todo.rb:315), which inserts into the `dependencies` table and updates the `updated_at` columns. This does not trigger `check_circular_dependencies`, which is only used run after `add_precessor` finishes. The error is shown in the UI, but the DB modification is not reverted.

**Comment by C-Otto** [comment]
This is the first time I used ruby, and I'm looking forward to your feedback!

**Comment by dnrce** [comment]
The changes from #1777 are in `master` but not `2.3_branch`.

**Comment by C-Otto** [comment]
Merge both #1777 and #1812 for `2.3_branch`, please.

**Comment by lrbalt** [comment]
I think I've got them all in 2.3 too

**Comment by C-Otto** [comment]
The UI issues are not solved in #1812, yet. I'd keep this issue open until the UI issues also are fixed.

**Comment by lrbalt** [comment]
ok, reopened

**Comment by C-Otto** [comment]
Please reopen (note to self: only mention fix messages in pull requests, not commits. And, close pull requests while they are not ready to merge).

**Comment by dnrce** [comment]
@C-Otto, in reference to your last comment above, absolutely do mention issue fixes in commit messages. Merges don't show up when doing a blame so it's very helpful to have the issue number in the actual commit message. The only problem here was that the fix was targeted against master when it should have been 2.3_branch.

**Comment by C-Otto** [comment]
OK, will do. 2.3_branch PR coming soon.

**Comment by dnrce** [comment]
Fixed in #1884.
